File: src/main/java/dev/foxgirl/mineseekdestroy/GameTeam.java
package dev.foxgirl.mineseekdestroy;

import dev.foxgirl.mineseekdestroy.util.collect.ImmutableList;
import net.minecraft.scoreboard.Scoreboard;
import net.minecraft.scoreboard.Team;
import net.minecraft.screen.ScreenTexts;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

public enum GameTeam {

    NONE(Text.of("NONE"), null, null, null, Formatting.WHITE, null),
    SKIP(Text.of("SKIP"), "msd_skip", "msd_skip_dead", null, Formatting.GREEN, Formatting.GRAY),
    GHOST(Text.of("GHOST"), "msd_ghost", "msd_ghost_dead", null, Formatting.GRAY, Formatting.GRAY),
    OPERATOR(Text.of("ADMIN"), "msd_operator", null, null, Formatting.GREEN, null),
    PLAYER_DUEL(Text.of("DUEL"), "msd_duel", "msd_duel_dead", "msd_duel_damaged", Formatting.RED, Formatting.DARK_GRAY),
    PLAYER_WARDEN(Text.of("WARDEN"), "msd_warden", "msd_warden_dead", "msd_warden_damaged", Formatting.RED, Formatting.DARK_RED),
    PLAYER_BLACK(Text.of("BLACK"), "msd_black", "msd_black_dead", "msd_black_damaged", Formatting.DARK_PURPLE, Formatting.DARK_GRAY),
    PLAYER_YELLOW(Text.of("YELLOW"), "msd_yellow", "msd_yellow_dead", "msd_yellow_damaged", Formatting.YELLOW, Formatting.GOLD),
    PLAYER_BLUE(Text.of("BLUE"), "msd_blue", "msd_blue_dead", "msd_blue_damaged", Formatting.AQUA, Formatting.BLUE);

    private final Text displayName;
    private final String nameAlive;
    private final String nameDead;
    private final String nameDamaged;
    private final Formatting colorAlive;
    private final Formatting colorDead;
    private final List<String> names;

    GameTeam(Text displayName, String nameAlive, String nameDead, String nameDamaged, Formatting colorAlive, Formatting colorDead) {
        this.displayName = displayName;
        this.nameAlive = nameAlive;
        this.nameDead = nameDead;
        this.nameDamaged = nameDamaged;
        this.colorAlive = colorAlive;
        this.colorDead = colorDead;

        var names = ImmutableList.<String>builder(3);
        if (nameAlive != null) names.add(nameAlive);
        if (nameDead != null) names.add(nameDead);
        if (nameDamaged != null) names.add(nameDamaged);
        this.names = names.build();
    }

    public boolean isPlaying() {
        return this != NONE && this != SKIP && this != GHOST && this != OPERATOR;
    }

    public boolean isOperator() {
        return this == OPERATOR;
    }

    public boolean isSpectator() {
        return this == NONE;
    }

    public boolean isGhost() {
        return this == SKIP || this == GHOST;
    }

    public boolean isPlayingOrGhost() {
        return this != NONE && this != OPERATOR;
    }

    public boolean isOnScoreboard() {
        return this != NONE && this != GHOST && this != OPERATOR;
    }

    public boolean isCannon() {
        return this == PLAYER_YELLOW || this == PLAYER_BLUE;
    }

    public @Nullable String getName() {
        return nameAlive;
    }

    public @NotNull List<@NotNull String> getNames() {
        return names;
    }

    public @NotNull Formatting getColor() {
        return colorAlive != null ? colorAlive : Formatting.WHITE;
    }

    public @NotNull Text getDisplayName() {
        return displayName != null ? displayName.copy().formatted(getColor()) : ScreenTexts.EMPTY;
    }

    private static Team getOrCreateTeam(Scoreboard scoreboard, String name, Consumer<Team> initializer) {
        var team = scoreboard.getTeam(name);
        if (team == null) {
            team = scoreboard.addTeam(name);
            initializer.accept(team);
        }
        return team;
    }

    public @Nullable Team getAliveTeam(@NotNull Scoreboard scoreboard) {
        Objects.requireNonNull(scoreboard, "Argument 'scoreboard'");
        if (nameAlive == null) return null;
        return getOrCreateTeam(scoreboard, nameAlive, team -> {
            team.setDisplayName(displayName);
            team.setColor(colorAlive);
        });
    }

    public @Nullable Team getDeadTeam(@NotNull Scoreboard scoreboard) {
        Objects.requireNonNull(scoreboard, "Argument 'scoreboard'");
        if (nameDead == null) return null;
        return getOrCreateTeam(scoreboard, nameDead, team -> {
            team.setDisplayName(displayName);
            team.setColor(colorDead);
            team.setPrefix(Text.of("\u2620 "));
        });
    }

    public @Nullable Team getDamagedTeam(@NotNull Scoreboard scoreboard) {
        Objects.requireNonNull(scoreboard, "Argument 'scoreboard'");
        if (nameDamaged == null) return null;
        return getOrCreateTeam(scoreboard, nameDamaged, team -> {
            team.setDisplayName(displayName);
            team.setColor(Formatting.LIGHT_PURPLE);
        });
    }

}

File: src/main/java/dev/foxgirl/mineseekdestroy/GameTeam.kt
package dev.foxgirl.mineseekdestroy

import dev.foxgirl.mineseekdestroy.util.collect.buildImmutableList
import net.minecraft.scoreboard.Scoreboard
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import net.minecraft.util.Formatting.*
import java.util.*
import net.minecraft.scoreboard.Team as ScoreboardTeam

enum class GameTeam(
    val color: Formatting = WHITE,
    val colorDead: Formatting = WHITE,
    private val hasTeam: Boolean = true,
    private val hasDeadTeam: Boolean = true,
    private val hasUndeadTeam: Boolean = true,
    private val hasDamagedTeam: Boolean = true,
) {

    // Definitions

    // Non-playing teams
    NONE(hasTeam = false),
    SKIP(color = GREEN, colorDead = GRAY, hasUndeadTeam = false, hasDamagedTeam = false),
    GHOST(color = GRAY, hasUndeadTeam = false, hasDamagedTeam = false),
    OPERATOR(color = GREEN, hasDeadTeam = false, hasUndeadTeam = false, hasDamagedTeam = false),
    // Actively playing teams
    DUELIST(color = RED, colorDead = DARK_GRAY, hasUndeadTeam = false),
    WARDEN(color = RED, colorDead = DARK_RED, hasUndeadTeam = false),
    BLACK(color = DARK_PURPLE, colorDead = DARK_GRAY, hasUndeadTeam = false),
    // Actively playing and canon teams
    YELLOW(color = Formatting.YELLOW, colorDead = Formatting.GOLD),
    BLUE(color = Formatting.AQUA, colorDead = Formatting.BLUE);

    // Implementation

    /** Is this player actively playing in the game? (not a ghost) */
    val isPlaying
        get() = this != NONE && this != SKIP && this != GHOST && this != OPERATOR
    /** Is this player actively playing in the game? (including ghosts) */
    val isPlayingOrGhost
        get() = this != NONE && this != OPERATOR
    /** Is this player a ghost? */
    val isGhost
        get() = this == SKIP || this == GHOST
    /** Is this player an operator? (on the operator team) */
    val isOperator
        get() = this == OPERATOR
    /** Is this player a spectator? (not a ghost) */
    val isSpectator
        get() = this == NONE
    /** Should this player be displayed on the scoreboard? */
    val isOnScoreboard
        get() = this != NONE && this != GHOST && this != OPERATOR
    /** Should this player be considered when automatically ending the round? */
    val isCanon
        get() = this == YELLOW || this == BLUE

    val displayName: Text = Text.literal(name).formatted(color)

    val teamName: String?
    val teamNameDead: String?
    val teamNameUndead: String?
    val teamNameDamaged: String?

    val teamNames: List<String>

    init {
        if (hasTeam) {
            teamName = "msd_${name.lowercase(Locale.ROOT)}"
            teamNameDead = if (hasDeadTeam) teamName + "_dead" else null
            teamNameUndead = if (hasUndeadTeam) teamName + "_undead" else null
            teamNameDamaged = if (hasDamagedTeam) teamName + "_damaged" else null
        } else {
            teamName = null
            teamNameDead = null
            teamNameUndead = null
            teamNameDamaged = null
        }
        teamNames = buildImmutableList(4) {
            if (teamName != null) add(teamName)
            if (teamNameDead != null) add(teamNameDead)
            if (teamNameUndead != null) add(teamNameUndead)
            if (teamNameDamaged != null) add(teamNameDamaged)
        }
    }

    private fun getOrCreateTeam(
        scoreboard: Scoreboard,
        name: String?, color: Formatting, block: (ScoreboardTeam) -> Unit = {},
    ): ScoreboardTeam? {
        if (name != null) {
            var team = scoreboard.getTeam(name)
            if (team == null) {
                team = scoreboard.addTeam(name)
                team.displayName = Text.literal(name).formatted(color)
                team.color = color
                block(team)
            }
            return team
        }
        return null
    }

    fun getTeam(scoreboard: Scoreboard)
        = getOrCreateTeam(scoreboard, teamName, color)
    fun getDeadTeam(scoreboard: Scoreboard)
        = getOrCreateTeam(scoreboard, teamNameDead, colorDead) { it.prefix = Text.of("\u2620 ") }
    fun getUndeadTeam(scoreboard: Scoreboard)
        = getOrCreateTeam(scoreboard, teamNameUndead, colorDead) { it.prefix = Text.of("\u2620 ") }
    fun getDamagedTeam(scoreboard: Scoreboard)
        = getOrCreateTeam(scoreboard, teamNameDamaged, LIGHT_PURPLE)

}
